File: leetcode-algorithms/138. Copy List with Random Pointe/Solution.cpp
// A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.
//
// Return a deep copy of the list.
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode* copyRandomList(RandomListNode* head) {
        if (head == nullptr) return head;

        map<RandomListNode*, RandomListNode*> nodemap;

        RandomListNode* p = head;
        while (p) {
            RandomListNode* node = new RandomListNode(p->label);
            nodemap[p] = node;
            p = p->next;
        }

        p = head;
        while (p) {
            nodemap[p]->next = nodemap[p->next];
            nodemap[p]->random = (p->random == nullptr) ? nullptr : nodemap[p->random];
            p = p->next;
        }

        return nodemap[head];
    }
};

File: leetcode-algorithms/138. Copy List with Random Pointe/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;
    public Node random;

    public Node() {}

    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
    public Node copyRandomList(Node head) {

        Map<Node, Node> vmap  = new HashMap<>();
        Node dummy = new Node(1, null, null);
        Node dp = dummy;
        Node cp = head;

        while (cp != null) {
            Node local = new Node(cp.val, null, null);
            dummy.next = local;
            dummy = local;
            vmap.put(cp, local);
            cp = cp.next;

        }

        cp = head;
        while (cp != null) {
            vmap.get(cp).random = (cp.random == null) ? null : vmap.get(cp.random);
            cp = cp.next;
        }

        return dp.next;
    }
}
